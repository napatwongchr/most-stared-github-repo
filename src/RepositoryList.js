import React, { useState, useEffect, useCallback } from "react";
import { css } from "emotion";
import { useAsync } from "react-async";
import axios from "axios";

function RepositoryList() {
  const [pageCount, setPageCount] = useState(1);
  const [isLoadMore, setIsLoadMore] = useState(false);
  const [repositories, setRepositories] = useState(null);

  const { isLoading } = useAsync({
    promiseFn: initRepository,
    setRepositories,
    setPageCount,
    page: pageCount
  });

  const { run: runLoadMoreRepository } = useAsync({
    deferFn: loadMoreRepository,
    setPageCount,
    setIsLoadMore
  });

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  useEffect(() => {
    if (!isLoadMore) return;
    runLoadMoreRepository(pageCount);
  }, [isLoadMore, pageCount, runLoadMoreRepository]);

  function handleScroll() {
    if (
      window.innerHeight + document.documentElement.scrollTop !==
      document.documentElement.offsetHeight
    ) {
      return;
    }
    setIsLoadMore(true);
  }

  if (isLoading) {
    return (
      <div
        className={css`
          font-size: 46px;
        `}
      >
        Loading ...
      </div>
    );
  }

  return (
    <div data-testid="repository-list">
      {repositories ? (
        repositories.items.map(repository => {
          return (
            <div
              key={repository.id}
              data-testid="repository-item"
              className={css`
                background-color: lightblue;
                margin-bottom: 25px;
                padding: 30px;
                &:hover {
                  cursor: pointer;
                }
              `}
              onClick={() => window.open(repository.html_url)}
            >
              <div
                className={css`
                  display: flex;
                  justify-content: space-between;
                `}
              >
                <span>
                  {repository.name} ({repository.language || "-"})
                </span>
                <div
                  className={css`
                    label: stats;
                  `}
                >
                  <span
                    className={css`
                      label: starCount;
                      margin-right: 30px;
                    `}
                  >
                    Stars: {repository.stargazers_count}
                  </span>
                  <span
                    className={css`
                      label: forkCount;
                    `}
                  >
                    Forks: {repository.forks_count}
                  </span>
                </div>
              </div>
              <div
                className={css`
                  label: description;
                  margin-top: 20px;
                `}
              >
                {repository.description}
              </div>
            </div>
          );
        })
      ) : (
        <div>No repositories</div>
      )}
    </div>
  );
}

async function initRepository({ setRepositories, setPageCount, page }) {
  const response = await searchRepository();
  setRepositories(response.data);
  setPageCount(page + 1);
}

async function loadMoreRepository(
  [pageCount],
  { setPageCount, setIsLoadMore }
) {
  setTimeout(() => {
    setIsLoadMore(false);
    setPageCount(pageCount + 1);
  }, 2000);
}

function searchRepository(
  order = "desc",
  page = 1,
  perPage = 10,
  createdAt,
  sort = "stars"
) {
  return axios.get(
    `https://api.github.com/search/repositories?q=created:>2019-09-12&sort=${sort}&order=${order}&page=${page}&per_page=${perPage}`,
    {
      auth: {
        username: "napatwongchr",
        password: "xJ3a8k8f2810023"
      }
    }
  );
}

export default RepositoryList;
