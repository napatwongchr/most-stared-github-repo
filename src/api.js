import axios from "axios";

export function searchRepository({ order, page, perPage, createdAt, sort }) {
  return axios.get(
    `https://api.github.com/search/repositories?q=created:>${createdAt}&sort=${sort}&order=${order}&page=${page}&per_page=${perPage}`,
    {
      auth: {
        username: "napatwongchr",
        password: "xJ3a8k8f2810023"
      }
    }
  );
}
