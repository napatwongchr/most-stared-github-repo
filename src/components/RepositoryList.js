import React, { useState, useEffect } from "react";
import { css, keyframes } from "emotion";
import { useAsync } from "react-async";
import * as api from "../api";

export default function RepositoryList() {
  const [pageCount, setPageCount] = useState(1);
  const [isLoadMore, setIsLoadMore] = useState(false);
  const [repositories, setRepositories] = useState(null);

  const { isLoading: isInitializeLoading, error: isInitializeError } = useAsync(
    {
      promiseFn: initRepository,
      setRepositories,
      setPageCount,
      page: pageCount
    }
  );

  const {
    isLoading: isLoadMoreLoading,
    error: isLoadMoreError,
    run: runLoadMoreRepository
  } = useAsync({
    deferFn: loadMoreRepository,
    setPageCount,
    setRepositories,
    setIsLoadMore
  });

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  useEffect(() => {
    if (!isLoadMore) return;
    runLoadMoreRepository(pageCount);
  }, [isLoadMore, pageCount, runLoadMoreRepository]);

  function handleScroll(event) {
    if (
      window.innerHeight + event.target.scrollingElement.scrollTop !==
      event.target.scrollingElement.offsetHeight
    ) {
      return;
    }
    setIsLoadMore(true);
  }

  if (isInitializeLoading) {
    return <div className={styles.spinnerContainer}>Loading ...</div>;
  }

  if (isInitializeError || isLoadMoreError) {
    return (
      <div className={styles.errorContainer}>Error fetching repositories.</div>
    );
  }

  return (
    <div className={styles.repoListContainer} data-testid="repository-list">
      {repositories ? (
        repositories.items.map(repository => {
          return (
            <div
              key={repository.id}
              className={styles.repoItemContainer}
              data-testid="repository-item"
              onClick={() => window.open(repository.html_url)}
            >
              <div className={styles.repoHeadingContainer}>
                <div className={styles.repoHeadingContent}>
                  <span className={styles.repoName}>{repository.name}</span>
                  {repository.language && (
                    <span className={styles.repoLanguageTag}>
                      {repository.language}
                    </span>
                  )}
                </div>
                <div className={styles.repoStatsContainer}>
                  <div className={styles.starsCountContianer}>
                    <span className={styles.starsCountText}>Stars:</span>{" "}
                    <span>{repository.stargazers_count}</span>
                  </div>
                  <div className={styles.forksCountContainer}>
                    <span className={styles.forksCountText}>Forks: </span>
                    <span>{repository.forks_count}</span>
                  </div>
                </div>
              </div>
              <div className={styles.repoDescContainer}>
                {repository.description}
              </div>
            </div>
          );
        })
      ) : (
        <div className={styles.noRepositoryContainer}>No repositories.</div>
      )}
      {isLoadMoreLoading && (
        <div className={styles.loadMoreRepositoryContainer}>
          <span>Loading more repositories...</span>
        </div>
      )}
    </div>
  );
}

const SEARCH_PARAMS = {
  order: "desc",
  page: 1,
  perPage: 10,
  query: `created:>${
    new Date(new Date().setMonth(new Date().getMonth() - 1))
      .toISOString()
      .split("T")[0]
  }`,
  sort: "stars"
};

async function initRepository({ setRepositories, setPageCount, page }) {
  const response = await api.searchRepository(SEARCH_PARAMS);
  setRepositories(response.data);
  setPageCount(page + 1);
}

async function loadMoreRepository(
  [pageCount],
  { setPageCount, setIsLoadMore, setRepositories }
) {
  const response = await api.searchRepository({
    ...SEARCH_PARAMS,
    page: pageCount
  });
  setIsLoadMore(false);
  setPageCount(pageCount + 1);
  setRepositories(prevRepositories => {
    return {
      ...prevRepositories,
      items: [...prevRepositories.items, ...response.data.items]
    };
  });
}

const animation = {
  aura: keyframes`
    0% {
      box-shadow: -2px 0px 20px -16px rgba(116, 94, 181, 1);
    }
    40% {
      box-shadow: -2px 0px 45px -16px rgba(116, 94, 181, 1);
    }
    70% {
      box-shadow: -2px 0px 56px -16px rgba(116, 94, 181, 1);
    }
    85% {
      box-shadow: -2px 0px 45px -16px rgba(116, 94, 181, 1);
    }
    100% {
      box-shadow: -2px 0px 20px -16px rgba(116, 94, 181, 1);
    }
  `
};

const styles = {
  spinnerContainer: css`
    display: flex;
    justify-content: center;
    font-size: 36px;
  `,
  errorContainer: css`
    display: flex;
    justify-content: center;
    font-size: 36px;
  `,
  noRepositoryContainer: css`
    display: flex;
    justify-content: center;
    font-size: 36px;
  `,
  repoListContainer: css`
    label: repoList;
  `,
  repoItemContainer: css`
    label: repoItem;
    background-color: #ffffff;
    margin-bottom: 40px;
    padding: 40px;
    border-top: 3px solid #745eb5;
    border-radius: 3px;
    &:hover {
      animation: ${animation.aura} 1.3s ease infinite;
      cursor: pointer;
    }
  `,
  repoHeadingContainer: css`
    display: flex;
    justify-content: space-between;
    @media (max-width: 576px) {
      display: flex;
      flex-direction: column;
    }
  `,
  repoHeadingContent: css`
    @media (max-width: 576px) {
      display: flex;
      flex-direction: column;
      align-items: center;
    }
  `,
  repoName: css`
    font-weight: bold;
    font-size: 22px;
    @media (max-width: 576px) {
      text-align: center;
    }
  `,
  repoLanguageTag: css`
    background-color: #745eb5;
    padding: 3px;
    margin-left: 5px;
    border-radius: 6px;
    color: #ffffff;
    @media (max-width: 576px) {
      margin-top: 20px;
    }
  `,
  repoStatsContainer: css`
    label: stats;
    display: flex;
    @media (max-width: 576px) {
      justify-content: center;
      margin-top: 30px;
    }
  `,
  starsCountContianer: css`
    label: starCount;
    margin-right: 30px;
  `,
  starsCountText: css`
    color: #f5a70a;
  `,
  forksCountContainer: css`
    label: forkCount;
  `,
  forksCountText: css`
    color: #00818a;
  `,
  repoDescContainer: css`
    label: description;
    margin-top: 30px;
    @media (max-width: 576px) {
      text-align: center;
    }
  `,
  loadMoreRepositoryContainer: css`
    display: flex;
    justify-content: center;
    margin-bottom: 40px;
  `
};
