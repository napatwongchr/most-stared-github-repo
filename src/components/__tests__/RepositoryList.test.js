import React from "react";
import {
  render,
  cleanup,
  waitFor,
  screen,
  fireEvent
} from "@testing-library/react";
import RepositoryList from "../RepositoryList";
import * as api from "../../api";

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});

test("Repository list should render properly", async () => {
  const apiCall = jest
    .spyOn(api, "searchRepository")
    .mockImplementationOnce(() => {
      return Promise.resolve({
        data: {
          items: [{ id: 1 }, { id: 2 }]
        }
      });
    });

  const { getByTestId } = render(<RepositoryList />);

  expect(screen.getByText(/loading/i)).toBeInTheDocument();
  await waitFor(() => expect(apiCall).toHaveBeenCalledTimes(1));

  expect(getByTestId("repository-list")).toBeInTheDocument();
  expect(getByTestId("repository-list").childNodes.length).toBe(2);
});

test("Repository list should show 'No repository found.' when no repository result", async () => {
  const apiCall = jest
    .spyOn(api, "searchRepository")
    .mockImplementationOnce(() => {
      return Promise.resolve({
        data: null
      });
    });

  render(<RepositoryList />);

  expect(screen.getByText(/loading/i)).toBeInTheDocument();

  await waitFor(() => expect(apiCall).toHaveBeenCalledTimes(1));

  expect(screen.getByText(/no repositories./i)).toBeInTheDocument();
});

test("Repository list should show 'Error fetching repositories.' when error occured", async () => {
  const apiCall = jest
    .spyOn(api, "searchRepository")
    .mockImplementationOnce(() => {
      return Promise.reject("error");
    });

  render(<RepositoryList />);

  expect(screen.getByText(/loading/i)).toBeInTheDocument();
  await waitFor(() => expect(apiCall).toHaveBeenCalledTimes(1));
  expect(screen.getByText(/error fetching repositories./i)).toBeInTheDocument();
});

test("Repository list should fetch when scroll to the end", async () => {
  const uuidv1 = require("uuid/v1");
  const apiCall = jest.spyOn(api, "searchRepository").mockImplementation(() => {
    return Promise.resolve({
      data: {
        items: [{ id: uuidv1() }, { id: uuidv1() }]
      }
    });
  });
  const { getByTestId } = render(<RepositoryList />);
  expect(screen.getByText(/loading/i)).toBeInTheDocument();
  await waitFor(() => expect(apiCall).toHaveBeenCalledTimes(1));

  fireEvent.scroll(window, {
    target: { scrollingElement: { scrollTop: 1000, offsetHeight: 1768 } }
  });

  expect(screen.getByText(/Loading more repositories.../i)).toBeInTheDocument();
  await waitFor(() => expect(apiCall).toHaveBeenCalledTimes(2));

  expect(getByTestId("repository-list").childNodes.length).toBe(4);
});

test("Repository item should contains information", async () => {
  const apiCall = jest.spyOn(api, "searchRepository").mockImplementation(() => {
    return Promise.resolve({
      data: {
        items: [
          {
            id: "1",
            name: "repo 1",
            description: "This is desc 1",
            stargazers_count: "100",
            forks_count: "10",
            language: "python"
          },
          {
            id: "2",
            name: "repo 2",
            description: "This is desc 2",
            stargazers_count: "50",
            forks_count: "5",
            language: "js"
          }
        ]
      }
    });
  });
  const { asFragment } = render(<RepositoryList />);
  expect(screen.getByText(/loading/i)).toBeInTheDocument();
  await waitFor(() => expect(apiCall).toHaveBeenCalledTimes(1));
  const firstRender = asFragment();
  expect(firstRender).toMatchSnapshot();
});
