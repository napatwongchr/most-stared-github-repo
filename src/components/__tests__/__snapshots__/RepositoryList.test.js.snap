// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository item should contains information 1`] = `
<DocumentFragment>
  <div
    class="css-1xgr2qs-repoList"
    data-testid="repository-list"
  >
    <div
      class="css-1vikx0g-repoItem"
      data-testid="repository-item"
    >
      <div
        class="css-mr42ks"
      >
        <div
          class="css-o7hb3"
        >
          <span
            class="css-mkpm6v"
          >
            repo 1
          </span>
          <span
            class="css-1180py9"
          >
            python
          </span>
        </div>
        <div
          class="css-1t511r-stats"
        >
          <div
            class="css-1jor1m6-starCount"
          >
            <span
              class="css-1u0c3ag"
            >
              Stars:
            </span>
             
            <span>
              100
            </span>
          </div>
          <div
            class="css-n6lfbi-forkCount"
          >
            <span
              class="css-cuwowd"
            >
              Forks: 
            </span>
            <span>
              10
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-1yuj63e-description"
      >
        This is desc 1
      </div>
    </div>
    <div
      class="css-1vikx0g-repoItem"
      data-testid="repository-item"
    >
      <div
        class="css-mr42ks"
      >
        <div
          class="css-o7hb3"
        >
          <span
            class="css-mkpm6v"
          >
            repo 2
          </span>
          <span
            class="css-1180py9"
          >
            js
          </span>
        </div>
        <div
          class="css-1t511r-stats"
        >
          <div
            class="css-1jor1m6-starCount"
          >
            <span
              class="css-1u0c3ag"
            >
              Stars:
            </span>
             
            <span>
              50
            </span>
          </div>
          <div
            class="css-n6lfbi-forkCount"
          >
            <span
              class="css-cuwowd"
            >
              Forks: 
            </span>
            <span>
              5
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-1yuj63e-description"
      >
        This is desc 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;
